sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala

```diff
@@ -381,6 +382,55 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
     }
   }
 
+  private def withPrunedPartSchema(
+      fss: GpuFileSourceScanExec,
+      referenceList: Seq[Expression]): GpuFileSourceScanExec = {
+    // Luckily partition columns do not support nested types. So only need to prune the
+    // top level columns.
+    val prunedPartSchema = StructType(fss.relation.partitionSchema.filter { f =>
+      // The partition columns used in the referenceList will be kept.
+      referenceList.exists { expr =>
+        expr.find {
+          // It is safe to use the column name for comparison for the two cases listed in
+          // `prunePartitionForFileSourceScan`
+          case attr: AttributeReference => attr.name == f.name
+          case _ => false
+        }.isDefined
+      }})
+    fss.copy(requiredPartitionSchema = Some(prunedPartSchema))(fss.rapidsConf)
+  }
+
+  // This tries to prune the partition schema for GpuFileSourceScanExec by leveraging
+  // the project list of the first GpuProjectExec after a GpuFileSourceScanExec.
+  private def prunePartitionForFileSourceScan(plan: SparkPlan): SparkPlan = plan match {
+    case p @ GpuProjectExec(projectList, fss: GpuFileSourceScanExec, _) =>
```

Comment:
This doesn't handle cases where the project or the filter cannot be placed on the GPU (e.g.: one of the expressions is a UDF or some other type of expression we do not support on the GPU).  I'm fine if this is tracked as a followup issue, as it's not a common occurrence.

Reply:
Here it is https://github.com/NVIDIA/spark-rapids/issues/7436.